paymentsheet/src/test/java/com/stripe/android/paymentelement/embedded/EmbeddedConfirmationHelperTest.kt

```diff
@@ -0,0 +1,128 @@
+import com.stripe.android.paymentelement.ExperimentalEmbeddedPaymentElementApi
+import com.stripe.android.paymentelement.confirmation.ConfirmationHandler
+import com.stripe.android.paymentelement.confirmation.FakeConfirmationHandler
+import com.stripe.android.paymentsheet.PaymentSheet
+import com.stripe.android.paymentsheet.model.PaymentSelection
+import com.stripe.android.paymentsheet.state.PaymentElementLoader
+import com.stripe.android.testing.CoroutineTestRule
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.test.runTest
+import org.junit.Rule
+import org.mockito.Mockito.mock
+import kotlin.test.Test
+
+@ExperimentalEmbeddedPaymentElementApi
+internal class EmbeddedConfirmationHelperTest {
+    @get:Rule
+    val coroutineTestRule = CoroutineTestRule()
+
+    @Test
+    fun constructorRegistersWithConfirmationHandler() = testScenario {
+        assertThat(confirmationHandler.registerTurbine.awaitItem()).isNotNull()
+    }
+
+    @Test
+    fun constructorWiresUpConfirmationHandlerToResultCallback() = testScenario {
+        assertThat(confirmationHandler.registerTurbine.awaitItem()).isNotNull()
+
+        val exception = IllegalStateException("Test failure.")
+        confirmationHandler.state.value = ConfirmationHandler.State.Complete(
+            ConfirmationHandler.Result.Failed(
+                cause = exception,
+                message = "Error".resolvableString,
+                type = ConfirmationHandler.Result.Failed.ErrorType.Internal,
+            )
+        )
+        assertThat(resultCallbackTurbine.awaitItem()).isEqualTo(
+            EmbeddedPaymentElement.Result.Failed(exception)
+        )
+    }
+
+    @Test
+    fun confirmCallsResultCallbackWithFailureWhenLoadedStateIsNull() = testScenario(
+        loadedState = null,
+    ) {
+        assertThat(confirmationHandler.registerTurbine.awaitItem()).isNotNull()
+        confirmationHelper.confirm()
+        assertThat(resultCallbackTurbine.awaitItem()).isInstanceOf<EmbeddedPaymentElement.Result.Failed>()
+    }
+
+    @Test
+    fun confirmCallsResultCallbackWithFailureWhenNoSelection() = testScenario(
+        loadedState = defaultLoadedState().copy(selection = null),
+    ) {
+        assertThat(confirmationHandler.registerTurbine.awaitItem()).isNotNull()
+        confirmationHelper.confirm()
+        assertThat(resultCallbackTurbine.awaitItem()).isInstanceOf<EmbeddedPaymentElement.Result.Failed>()
+    }
+
+    @Test
+    fun confirmCallsConfirmationHandlerStart() = testScenario {
+        assertThat(confirmationHandler.registerTurbine.awaitItem()).isNotNull()
+        confirmationHelper.confirm()
+        assertThat(confirmationHandler.startTurbine.awaitItem()).isNotNull()
+    }
+
+    private fun defaultLoadedState(): EmbeddedConfirmationHelper.State {
+        return EmbeddedConfirmationHelper.State(
+            paymentMethodMetadata = PaymentMethodMetadataFactory.create(),
+            selection = PaymentSelection.GooglePay,
```

Comment:
I chose this because it had the least amount of setup required to use it. But happy to choose something else if it makes more sense.

Reply:
This is fine, I don't think the selection matters here, just proper interactions.